Added support for filesystem metrics on Docker

JohnnyG235711 opened the pull request:
First reported in issue #2609 (https://github.com/google/cadvisor/issues/2609)

JohnnyG235711 commented:
@googlebot I signed it!

iwankgb commented on fs/fs.go:
```diff
@@ -483,6 +483,17 @@ func getDiskStatsMap(diskStatsFile string) (map[string]DiskStats, error) {
 			}
 		}
+
+		major64, error :=  strconv.ParseUint(words[0], 10, 64)
```
Use `err` instead of `error` as error-storing variable, please.

JohnnyG235711 replied:
Sure, I have renamed to "err" in my latest commit.

iwankgb commented on fs/fs.go:
```diff
@@ -439,7 +439,7 @@ func (i *RealFsInfo) GetFsInfoForPath(mountSet map[string]struct{}) ([]Fs, error
 var partitionRegex = regexp.MustCompile(`^(?:(?:s|v|xv)d[a-z]+\d*|dm-\d+)$`)
 
-func getDiskStatsMap(diskStatsFile string) (map[string]DiskStats, error) {
+func GetDiskStatsMap(diskStatsFile string) (map[string]DiskStats, error) {
```
Do you think it would be possible to use `RealFsInfo.GetFsInfoForPath()` rather than exporting yet another function? It would be great to limit public interfaces as much as possible.

JohnnyG235711 replied:
It was indeed possible and I have made getDiskStatsMap private again in my latest commit.

iwankgb reviewed (changes requested):
@JohnnyG235711 you will have to sign CLA before this PR can be merged. I would appreciate taking look at my comments too :)

JohnnyG235711 commented:
@iwankgb Thanks for taking the time to look at this. I have changed my code to be inline with your code review comments and pushed the changes into my branch. Hopefully you can see them for the re-review. I have also updated my primary email address to match the one in the google CLA corporate agreement for my company, which is Syncsort Incorporated. Thanks.

iwankgb commented:
@JohnnyG235711 you probably need to talk to someone at your company to be added to an AD (or similar) group. The bot still claims you have not signed the CLA.

JohnnyG235711 commented:
@googlebot I fixed it.

JohnnyG235711 commented:
> @JohnnyG235711 you probably need to talk to someone at your company to be added to an AD (or similar) group. The bot still claims you have not signed the CLA.

@iwankgb I'm in the group now, but now getting a different error: "We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors."
There's only one author though, and it's me. I'm wondering if it's because I didn't set my email in the previous commits or maybe because I merged upwards. If so maybe I need to try again with the different branch.

iwankgb commented:
@JohnnyG235711 if I were you I would try to rebase against master and then squash all the commits. It might help.

JohnnyG235711 commented:
@iwankgb Thank you, that worked.

iwankgb commented:
/ok-to-test

iwankgb reviewed (changes requested):
It segfaults on my RPi 4:
```
I0110 16:38:12.163674   88322 manager.go:987] Added container: "/user.slice" (aliases: [], namespace: "")
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x9fc39c]

goroutine 56 [running]:
github.com/google/cadvisor/container/docker.(*dockerContainerHandler).addDiskStats(0x40000dd2c0, 0x400086e7a8, 0x0)
	/home/ubuntu/go/src/github.com/google/cadvisor/container/docker/handler.go:430 +0xa4
github.com/google/cadvisor/container/docker.(*dockerContainerHandler).getFsStats(0x40000dd2c0, 0x4000805800, 0x0, 0x0)
	/home/ubuntu/go/src/github.com/google/cadvisor/container/docker/handler.go:418 +0x23c
github.com/google/cadvisor/container/docker.(*dockerContainerHandler).GetStats(0x40000dd2c0, 0x0, 0x0, 0x0)
	/home/ubuntu/go/src/github.com/google/cadvisor/container/docker/handler.go:465 +0x68
github.com/google/cadvisor/manager.(*containerData).updateStats(0x4000788000, 0xbff6e93d09287079, 0xace62627)
	/home/ubuntu/go/src/github.com/google/cadvisor/manager/container.go:637 +0x54
github.com/google/cadvisor/manager.(*containerData).housekeepingTick(0x4000788000, 0x400070a2a0, 0x5f5e100, 0x4000127100)
	/home/ubuntu/go/src/github.com/google/cadvisor/manager/container.go:583 +0xf4
github.com/google/cadvisor/manager.(*containerData).housekeeping(0x4000788000)
	/home/ubuntu/go/src/github.com/google/cadvisor/manager/container.go:531 +0x1bc
created by github.com/google/cadvisor/manager.(*containerData).Start
	/home/ubuntu/go/src/github.com/google/cadvisor/manager/container.go:119 +0x3c
ubuntu@malyna4:~/go/src/github.com/google/cadvisor$ uname -a
Linux malyna4 5.4.0-1025-raspi #28-Ubuntu SMP PREEMPT Wed Dec 9 17:10:53 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux
ubuntu@malyna4:~/go/src/github.com/google/cadvisor$ go version
go version go1.13.8 linux/arm64
```

iwankgb commented:
I think that we are missing an important test here: CI should not pass :/

JohnnyG235711 commented:
I have added a failing test, which I then fixed by fixing the bug.

JohnnyG235711 commented:
@iwankgb Is my revised code OK?

iwankgb commented:
@JohnnyG235711 sorry for the delay, it's been hectic time at work. I will take another look at your code this week.

JohnnyG235711 commented:
@iwankgb No worries, and thanks for looking at it.

Creatone commented on fs/fs_test.go:
```diff
@@ -44,7 +44,7 @@ func TestMountInfoFromDir(t *testing.T) {
 	diskStatsMap, err := getDiskStatsMap("test_resources/diskstats")
 	if err != nil {
-		t.Errorf("Error calling getDiskStatMap %s", err)
+		t.Errorf("Error calling GetDiskStatMap %s", err)
```
The function didn't change the name, so in my opinion, you shouldn't change this error.

Creatone reviewed (commented):
Few comments.

JohnnyG235711 commented:
@Creatone Thanks for your feedback. I have amended the code in line with your review. Could you re-review?

Creatone commented:
@JohnnyG235711 please squash commits :)

Creatone reviewed (approved):
LGTM

JohnnyG235711 commented:
@Createone Thanks for looking at it
@iwankgb Looks like before merging can occur you also need to approve it. I believe that I have fixed the seg fault.

iwankgb commented:
LGTM

JohnnyG235711 commented:
@bobbypage   I was wondering when this might be tested and then merged? Thanks very much.

bobbypage commented:
LGTM, thanks.
